File: scripts/extract-country-from-optimized.js
#!/usr/bin/env node
/**
 * 從 hkn 項目的優化 GADM 文件中提取單個國家的數據
 * 
 * 用法:
 *   node scripts/extract-country-from-optimized.js TWN
 *   node scripts/extract-country-from-optimized.js GBR
 */

const fs = require('fs');
const path = require('path');

const countryCode = process.argv[2];
const level = process.argv[3] || '0'; // 默認 Level 0

if (!countryCode) {
  console.error('用法: node scripts/extract-country-from-optimized.js <COUNTRY_CODE> [LEVEL]');
  console.error('例如: node scripts/extract-country-from-optimized.js TWN 0');
  process.exit(1);
}

// 優先檢查多個可能的數據源位置
const possiblePaths = [
  path.join(__dirname, '../../map/data/gadm/gadm_level' + level + '.geojson'),
  path.join(__dirname, '../../hkn/data/gadm/optimized/gadm_level' + level + '_optimized.geojson'),
  '/Users/yulincho/Documents/01_Github/map/data/gadm/gadm_level' + level + '.geojson'
];

// 找到第一個存在的文件
let sourceFile = null;
for (const filePath of possiblePaths) {
  if (fs.existsSync(filePath)) {
    sourceFile = filePath;
    console.log(`✅ 找到數據文件: ${sourceFile}`);
    break;
  }
}
const targetDir = path.join(__dirname, '../../data/gadm', countryCode);
const targetFile = path.join(targetDir, `${countryCode}_${level}.geojson`);

if (!sourceFile || !fs.existsSync(sourceFile)) {
  console.error(`錯誤: 源文件不存在`);
  console.error(`檢查的路徑:`);
  console.error(`  1. ${path.join(mapDataPath, `gadm_level${level}.geojson`)}`);
  console.error(`  2. ${path.join(hknDataPath, `gadm_level${level}_optimized.geojson`)}`);
  process.exit(1);
}

console.log(`正在從 ${sourceFile} 提取 ${countryCode} (Level ${level})...`);

try {
  // 讀取源文件
  const sourceData = JSON.parse(fs.readFileSync(sourceFile, 'utf8'));
  
  // 過濾出指定國家的要素
  const filteredFeatures = sourceData.features.filter(feature => {
    const props = feature.properties || {};
    // GADM 屬性名稱可能是 GID_0, ISO, COUNTRY_CODE 等
    const gid0 = props.GID_0 || props.ISO || props.COUNTRY_CODE || props.GID_0;
    return gid0 === countryCode;
  });

  if (filteredFeatures.length === 0) {
    console.warn(`警告: 在源文件中未找到 ${countryCode} 的數據`);
    console.log('可用的國家代碼 (前10個):');
    const countries = new Set();
    sourceData.features.slice(0, 100).forEach(f => {
      const gid0 = f.properties?.GID_0 || f.properties?.ISO;
      if (gid0) countries.add(gid0);
    });
    console.log(Array.from(countries).slice(0, 10).join(', '));
    process.exit(1);
  }

  // 創建目標目錄
  if (!fs.existsSync(targetDir)) {
    fs.mkdirSync(targetDir, { recursive: true });
  }

  // 創建新的 GeoJSON
  const countryData = {
    type: 'FeatureCollection',
    features: filteredFeatures
  };

  // 寫入文件
  fs.writeFileSync(targetFile, JSON.stringify(countryData, null, 2));
  
  console.log(`✅ 成功提取 ${filteredFeatures.length} 個要素到 ${targetFile}`);
  const fileSize = fs.statSync(targetFile).size;
  console.log(`   文件大小: ${(fileSize / 1024 / 1024).toFixed(2)} MB`);
  
} catch (error) {
  console.error('錯誤:', error.message);
  process.exit(1);
}

File: scripts/extract-country-from-optimized.py
#!/usr/bin/env python3
"""
從 GADM 文件中提取單個國家的數據（支持大文件流式處理）

用法:
    python3 scripts/extract-country-from-optimized.py TWN 0
    python3 scripts/extract-country-from-optimized.py GBR 0
"""

import json
import sys
import os
from pathlib import Path

def find_source_file(level):
    """查找源數據文件"""
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    
    possible_paths = [
        project_root / 'map' / 'data' / 'gadm' / f'gadm_level{level}.geojson',
        project_root / 'hkn' / 'data' / 'gadm' / 'optimized' / f'gadm_level{level}_optimized.geojson',
        Path('/Users/yulincho/Documents/01_Github/map/data/gadm/gadm_level{level}.geojson'.format(level=level))
    ]
    
    for path in possible_paths:
        if path.exists():
            return path
    
    return None

def extract_country_streaming(source_file, country_code, target_file):
    """使用流式處理提取國家數據（適用於大文件）"""
    print(f"正在從 {source_file} 提取 {country_code}...")
    print(f"（使用流式處理以處理大文件）")
    
    features_found = 0
    features_written = 0
    
    # 使用 ijson 進行流式處理（如果可用），否則使用標準方法
    try:
        import ijson
        use_ijson = True
    except ImportError:
        use_ijson = False
        print("提示: 安裝 ijson 可以更快處理大文件: pip install ijson")
    
    if use_ijson:
        # 使用 ijson 流式處理
        with open(source_file, 'rb') as f, open(target_file, 'w', encoding='utf-8') as out:
            out.write('{"type":"FeatureCollection","features":[')
            first = True
            
            parser = ijson.items(f, 'features.item')
            for feature in parser:
                features_found += 1
                props = feature.get('properties', {})
                gid0 = props.get('GID_0') or props.get('ISO') or props.get('COUNTRY_CODE')
                
                if gid0 == country_code:
                    if not first:
                        out.write(',')
                    json.dump(feature, out, ensure_ascii=False)
                    features_written += 1
                    first = False
                    if features_written % 10 == 0:
                        print(f"  已提取 {features_written} 個要素...", end='\r')
            
            out.write(']}')
    else:
        # 標準方法：分塊讀取（對於超大文件可能仍會失敗）
        print("使用標準 JSON 解析（文件較大時可能需要較長時間）...")
        with open(source_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        features = data.get('features', [])
        print(f"總共找到 {len(features)} 個要素，正在過濾...")
        
        filtered_features = []
        for feature in features:
            props = feature.get('properties', {})
            gid0 = props.get('GID_0') or props.get('ISO') or props.get('COUNTRY_CODE')
            if gid0 == country_code:
                filtered_features.append(feature)
            features_found += 1
            if features_found % 1000 == 0:
                print(f"  已處理 {features_found}/{len(features)} 個要素...", end='\r')
        
        features_written = len(filtered_features)
        country_data = {
            'type': 'FeatureCollection',
            'features': filtered_features
        }
        
        with open(target_file, 'w', encoding='utf-8') as out:
            json.dump(country_data, out, ensure_ascii=False, indent=2)
    
    return features_written

def main():
    if len(sys.argv) < 2:
        print("用法: python3 scripts/extract-country-from-optimized.py <COUNTRY_CODE> [LEVEL]")
        print("例如: python3 scripts/extract-country-from-optimized.py TWN 0")
        sys.exit(1)
    
    country_code = sys.argv[1].upper()
    level = sys.argv[2] if len(sys.argv) > 2 else '0'
    
    # 查找源文件
    source_file = find_source_file(level)
    if not source_file:
        print("錯誤: 找不到源數據文件")
        print("檢查的路徑:")
        script_dir = Path(__file__).parent
        project_root = script_dir.parent
        print(f"  1. {project_root / 'map' / 'data' / 'gadm' / f'gadm_level{level}.geojson'}")
        print(f"  2. {project_root / 'hkn' / 'data' / 'gadm' / 'optimized' / f'gadm_level{level}_optimized.geojson'}")
        print(f"  3. /Users/yulincho/Documents/01_Github/map/data/gadm/gadm_level{level}.geojson")
        sys.exit(1)
    
    # 創建目標目錄和文件
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    target_dir = project_root / 'data' / 'gadm' / country_code
    target_file = target_dir / f'{country_code}_{level}.geojson'
    
    target_dir.mkdir(parents=True, exist_ok=True)
    
    # 提取數據
    try:
        features_count = extract_country_streaming(source_file, country_code, target_file)
        
        if features_count == 0:
            print(f"\n警告: 在源文件中未找到 {country_code} 的數據")
            # 顯示一些可用的國家代碼
            print("\n嘗試讀取前幾個要素以顯示可用的國家代碼...")
            try:
                with open(source_file, 'r', encoding='utf-8') as f:
                    # 只讀取開頭部分來獲取樣本
                    sample = f.read(50000)  # 讀取前 50KB
                    if '"GID_0"' in sample or '"ISO"' in sample:
                        print("（文件太大，無法快速顯示所有國家代碼）")
            except:
                pass
            sys.exit(1)
        
        file_size = os.path.getsize(target_file)
        print(f"\n✅ 成功提取 {features_count} 個要素到 {target_file}")
        print(f"   文件大小: {file_size / 1024 / 1024:.2f} MB")
        
    except Exception as e:
        print(f"\n錯誤: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == '__main__':
    main()




